### ISSUE TITLE
[PYD-136] ModelAfterValidator causing warnings from Pylance and mypy

### ISSUE THREAD

**Issue by Lithimlin** [pydantic/pydantic#6277]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2 installed directly from the `main` branch, or equivalent

### Description

I want to use an ModelAvterValidator via the `model_validator` decorator, however, both Pylance and mypy are issuing warnings when doing so. To make sure I was using it correctly, I tried using the example code from the documentation and I am experiencing the same issues there.
The code below results in the following two warnings:
```
Argument 1 has incompatible type "Callable[[UserModel, UserModel], Any]"; expected "Union[ModelAfterValidator, ModelAfterValidatorWithoutInfo]"  [arg-type] (mypy)

Argument of type "(cls: Self@UserModel, m: UserModel) -> UserModel" cannot be assigned to parameter of type "_AnyModeAfterValidator"
  Type "(cls: Self@UserModel, m: UserModel) -> UserModel" cannot be assigned to type "_AnyModeAfterValidator"
    Type "(cls: Self@UserModel, m: UserModel) -> UserModel" cannot be assigned to type "(self: _ModelType@__call__, __info: ValidationInfo) -> _ModelType@__call__"
      Parameter name mismatch: "self" versus "cls"
      Parameter 2: type "ValidationInfo" cannot be assigned to type "UserModel"
        "ValidationInfo" is incompatible with "UserModel"
    Type "(cls: Self@UserModel, m: UserModel) -> UserModel" cannot be assigned to type "(self: _ModelType@__call__) -> _ModelType@__call__"
      Parameter name mismatch: "self" versus "cls" (pylance)
```

When changing the parameter name `cls` to `self` as suggested by Pylance, the corresponding lines in the latter warning disappear, but nothing else happens and I am not sure whether that is what I am supposed to do there.

As my IDE I use VS Code OSS with the python, pylance, and mypy extensions, among others.

### Example Code

```Python
from pydantic import BaseModel, model_validator


class UserModel(BaseModel):
    username: str
    password1: str
    password2: str

    @model_validator(mode="after")
    def check_passwords_match(cls, m: "UserModel"):
        pw1 = m.password1
        pw2 = m.password2
        if pw1 is not None and pw2 is not None and pw1 != pw2:
            raise ValueError("passwords do not match")
        return m
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.0b3
pydantic-core version: 0.39.0 release build profile
install path: <path-to-workspace>/env/lib/python3.11/site-packages/pydantic
python version: 3.11.3 (main, Jun  5 2023, 09:32:32) [GCC 13.1.1 20230429]
platform: Linux-6.3.9-arch1-1-x86_64-with-glibc2.37
optional deps. installed: ['typing-extensions']
```

Selected Assignee: @hramezani

Selected Assignee: @Kludex

<sub>[PYD-136](https://linear.app/pydantic/issue/PYD-136/modelaftervalidator-causing-warnings-from-pylance-and-mypy)</sub>

Selected Assignee: @lig

**Comment by Kludex** [comment]
You can add the decorator `@classmethod`, and the IDE will be happy. 🙏

We've already documented this behavior.

**Comment by Lithimlin** [comment]
Thank you, it seems I missed this documentation then.

**Comment by Viicos** [comment]
@Kludex even with the `@classmethod` decorator, `mypy` is still unhappy:

```python
from pydantic import BaseModel, model_validator


class UserModel(BaseModel):
    username: str
    password1: str
    password2: str

    @model_validator(mode='after')
    @classmethod
    def check_passwords_match(cls, m: 'UserModel'):
        pw1 = m.password1
        pw2 = m.password2
        if pw1 is not None and pw2 is not None and pw1 != pw2:
            raise ValueError('passwords do not match')
        return m
```

```sh
error: Argument 1 has incompatible type "Callable[[type[UserModel], UserModel], Any]"; expected "ModelAfterValidator | ModelAfterValidatorWithoutInfo"  [arg-type]
```

(example taken from [docs](https://docs.pydantic.dev/dev-v2/usage/validators/#model-validators), `mypy 1.4.1`)

Looking at the excepted signatures:

https://github.com/pydantic/pydantic/blob/e0c0fe8cd6b6a1a5fdd43ab92d9814b278e324ad/pydantic/functional_validators.py#L402-L422

They doesn't seem to match the one used in the example

**Comment by Kludex** [comment]
Right. My bad. I'm going to check properly.

**Comment by adriangb** [comment]
Can you use this instead?

```py
from pydantic import BaseModel, model_validator


class UserModel(BaseModel):
    username: str
    password1: str
    password2: str

    @model_validator(mode="after")
    def check_passwords_match(self) -> "UserModel":
        if self.password1 != self.password2:
            raise ValueError("passwords do not match")
        return self
```

**Comment by adriangb** [comment]
It seems like this works with pyright but not mypy. @Kludex  if you didn't already try the above, can you look into why it works with pyright but not mypy? It might be worth asking on the mypy issue tracker.

**Comment by Cirn09** [comment]
> Can you use this instead?
> 
> ```python
> from pydantic import BaseModel, model_validator
> 
> 
> class UserModel(BaseModel):
>     username: str
>     password1: str
>     password2: str
> 
>     @model_validator(mode="after")
>     def check_passwords_match(self) -> "UserModel":
>         if self.password1 != self.password2:
>             raise ValueError("passwords do not match")
>         return self
> ```

no warning with pyright.

**Comment by treykeown** [comment]
+1, played around with this for a bit but can't get any model validators to work with mypy on V2

**Comment by zakstucke** [comment]
+1, @adriangb your example fails for me with mypy '1.4.0' with 

```
Argument 1 has incompatible type "Callable[[type[UserModel]], UserModel]"; expected "ModelAfterValidator | ModelAfterValidatorWithoutInfo"  [arg-type]
```

Note only `mode="after"` i've found problematic, `mode="before"` seems to work perfectly :1st_place_medal:

**Comment by Kludex** [comment]
Reported on mypy for now: https://github.com/python/mypy/issues/15620

**Comment by adriangb** [comment]
I think https://github.com/pydantic/pydantic/pull/6514 will fix this

**Comment by zakstucke** [comment]
@adriangb I think this is still acting up after #6514. 

I'm running `main` branch and mypy `1.4.1`. See reproduction below.

```Python
import typing as tp
from pydantic import BaseModel, model_validator


class Foo(BaseModel):
    foo: str

    def bar(self):
        return self.foo

    @model_validator(mode="after")
    def val(self):
        print(
            self.bar()
        )  # ERR Missing positional argument "self" in call to "bar"  [call-arg] (mypy)

        tp.reveal_type(self)
        # Revealed type is "Type[backend.mypy_playground.Foo]" (mypy)

        return self


Foo(foo="FOO")

# Correct at runtime,
# prints "FOO"
```

**Comment by Kludex** [comment]
I can't reproduce it with main, and mypy 1.4.1. It works properly here.

**Comment by zakstucke** [comment]
My bad, looks like my mypy cache was playing up, sorry.

Tested in a clean repo and seems to be working!

**Comment by daniil-berg** [comment]
Unfortunately, usage of `@model_validator(mode="after")` is still causing Mypy issues, even after #6514.

```python
from pydantic import BaseModel, model_validator


class Foo(BaseModel):
    x: int

    @model_validator(mode="after")  # error: Cannot infer function type argument  [misc]
    def print_x(self) -> "Foo":
        print(f"{self.x=}")
        return self  # error: Incompatible return value type (got "type[Foo]", expected "Foo")  [return-value]


Foo(x=1)  # prints "self.x=1"
```

Version info:

```
             pydantic version: 2.0.3
        pydantic-core version: 2.3.0 release build profile
                 install path: /path/to/.venv/lib/python3.9/site-packages/pydantic
               python version: 3.9.17 (main, Jun 24 2023, 20:40:40)  [GCC 13.1.1 20230429]
                     platform: Linux-6.4.3-arch1-1-x86_64-with-glibc2.37
     optional deps. installed: ['typing-extensions']
```

Also reproduced with:
```
               python version: 3.11.4 (main, Jun 24 2023, 20:38:45) [GCC 13.1.1 20230429]
```

Tested both with Mypy `1.4.1`.

**Comment by adriangb** [comment]
It must be an issue with your mypy config. I just copied your example to a `test.py` file and ran it with both mypy 1.1.1 and 1.4.1:

```
pydantic on  validators [$?] via 🐍 v3.10.9 (.venv) on ☁️  adrian@adriangb.com 
❯ mypy -V        
mypy 1.1.1 (compiled: yes)

pydantic on  validators [$?] via 🐍 v3.10.9 (.venv) on ☁️  adrian@adriangb.com 
❯ mypy test.py
Success: no issues found in 1 source file

pydantic on  validators [$?] via 🐍 v3.10.9 (.venv) on ☁️  adrian@adriangb.com 
❯ pip install -U mypy                                                                                            
Requirement already satisfied: mypy in ./.venv/lib/python3.10/site-packages (1.1.1)
...
Successfully installed mypy-1.4.1

pydantic on  validators [$?] via 🐍 v3.10.9 (.venv) on ☁️  adrian@adriangb.com 
❯ mypy test.py       
Success: no issues found in 1 source file

pydantic on  validators [$?] via 🐍 v3.10.9 (.venv) on ☁️  adrian@adriangb.com took 4s
```

**Comment by daniil-berg** [comment]
@adriangb Ah, good call! It turns out that the errors only come up in conjunction with the [Pydantic Mypy plugin](https://docs.pydantic.dev/latest/integrations/mypy/#enabling-the-plugin). I deactivated it and the errors disappeared. Can you reproduce them, if you enable the plugin? If so, should I open a separate issue for this?

**Comment by adriangb** [comment]
Yes please open a separate issue!
